File: LeetCode/Medium/0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    int rowLen;
    int colLen;
    public void solve(char[][] board) {
        rowLen = board.length;
        colLen = board[0].length;
        for (int i = 0; i < rowLen; ++i){
            if (board[i][0] == 'O') dfs(i, 0, board);
            if (board[i][colLen - 1] == 'O') dfs(i, colLen - 1, board);
        }
        
        for (int i = 1; i < colLen - 1; ++i){
            if (board[0][i] == 'O') dfs(0, i, board);
            if (board[rowLen - 1][i] == 'O') dfs(rowLen - 1, i, board);
        }
        
        for (int row = 0; row < rowLen; ++row){
            for (int col = 0; col < colLen; ++col){
                if (board[row][col] == 'T') board[row][col] = 'O';
                else if (board[row][col] == 'O') board[row][col] = 'X';
            }
        }
    }
    
    private void dfs(int row, int col, char[][] board){
        if (row < 0 || row == rowLen || col < 0 || col == colLen || board[row][col] != 'O') return;
        
        board[row][col] = 'T';
        
        dfs(row - 1, col, board);
        dfs(row + 1, col, board);
        dfs(row, col - 1, board);
        dfs(row, col + 1, board);
    }
}

File: LeetCode/Medium/0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rowLen = len(board)
        colLen = len(board[0])

        def dfs(row, col):
            if row < 0 or col < 0 or row == rowLen or col == colLen or board[row][col] != "O":
                return
            board[row][col] = "T"

            dfs(row - 1, col)
            dfs(row + 1, col)
            dfs(row, col - 1)
            dfs(row, col + 1)
        
        for i in range(rowLen):
            if board[i][0] == "O":
                dfs(i, 0)
            if board[i][colLen-1] == "O":
                dfs(i, colLen-1)
        
        for i in range(colLen):
            if board[0][i] == "O":
                dfs(0, i)
            if board[rowLen - 1][i] == "O":
                dfs(rowLen-1, i)
        
        for row in range(rowLen):
            for col in range(colLen):
                if board[row][col] == "T":
                    board[row][col] = "O"
                elif board[row][col] == "O":
                    board[row][col] = "X"